gtsam/hybrid/tests/testHybridEstimation.cpp

```diff
@@ -432,6 +432,69 @@ TEST(HybridEstimation, ProbabilityMultifrontal) {
   EXPECT(assert_equal(discrete_seq, hybrid_values.discrete()));
 }
 
+/****************************************************************************/
+/**
+ * Test for correctness via sampling.
+ *
+ * Given the conditional P(x0, m0, x1| z0, z1)
+ * with meaasurements z0, z1, we:
+ * 1. Start with the corresponding Factor Graph `FG`.
+ * 2. Eliminate the factor graph into a Bayes Net `BN`.
+ * 3. Sample from the Bayes Net.
+ * 4. Check that the ratio `BN(x)/FG(x) = constant` for all samples `x`.
+ */
+TEST(HybridEstimation, CorrectnessViaSampling) {
+  HybridNonlinearFactorGraph nfg;
+
+  auto noise_model = noiseModel::Diagonal::Sigmas(Vector1(1.0));
+  auto zero_motion =
+      boost::make_shared<BetweenFactor<double>>(X(0), X(1), 0, noise_model);
+  auto one_motion =
+      boost::make_shared<BetweenFactor<double>>(X(0), X(1), 1, noise_model);
+  std::vector<NonlinearFactor::shared_ptr> factors = {zero_motion, one_motion};
+  nfg.emplace_nonlinear<PriorFactor<double>>(X(0), 0.0, noise_model);
+  nfg.emplace_hybrid<MixtureFactor>(
+      KeyVector{X(0), X(1)}, DiscreteKeys{DiscreteKey(M(0), 2)}, factors);
+
+  Values initial;
+  double z0 = 0.0, z1 = 1.0;
+  initial.insert<double>(X(0), z0);
+  initial.insert<double>(X(1), z1);
+
+  // 1. Create the factor graph from the nonlinear factor graph.
+  HybridGaussianFactorGraph::shared_ptr fg = nfg.linearize(initial);
+  // 2. Eliminate into BN
+  Ordering ordering = fg->getHybridOrdering();
+  HybridBayesNet::shared_ptr bn = fg->eliminateSequential(ordering);
+
+  // Set up sampling
+  std::random_device rd;
+  std::mt19937_64 gen(11);
+
+  // 3. Do sampling
+  std::vector<double> ratios;
+  int num_samples = 1000;
+
+  for (size_t i = 0; i < num_samples; i++) {
+    // Sample from the bayes net
+    HybridValues sample = bn->sample(&gen);
+
+    // Compute the ratio in log form and canonical form
+    DiscreteValues assignment = sample.discrete();
+    double log_ratio = bn->error(sample.continuous(), assignment) -
+                       fg->error(sample.continuous(), assignment);
+    double ratio = exp(-log_ratio);
+
+    // Store the ratio for post-processing
+    ratios.push_back(ratio);
+  }
+
+  // 4. Check that all samples == 1.0 (constant)
```

Comment:
This is too clever for me. Explain in comment

Reply:
Why not just do the EXPECT_DOUBLES_EQUAL in the loop above? Simpler than to check the sum - which also might accidentally sum to 1. (very unlikely I know). You then also don't have to store the ratios.

Reply:
Actually, it is also wrong: ratios should be the same, not 1. Advice to make a lambda to compute log-ratio, then call it once to establish desired ratio, then 10 times to check.

Reply:
FYI model selection with the normalizing term is not in this PR. There's another open PR for that. That is why we're seeing the ratio as 1 instead of the determinant value.

I'll retarget that PR to this one and update the model selection code (aka change max product to sum product in elimination). That should give us the desired results.

Reply:
I added a lambda `compute_ratio` to compute the ratio both for the singular time as well as for `num_samples=10` samples.